### PR TITLE
ftp: Improved FTP active mode handling

### PR DESCRIPTION
**Pull request by jlucovsky**
Link to redmine ticket:

https://redmine.openinfosecfoundation.org/issues/2459
https://redmine.openinfosecfoundation.org/issues/2527
Describe changes:

- Detect FTP active mode over IPv4 and IPv6
- Create expectation based on mode
- Remove return type from TCP, ipv4 options and update tests
- Improve unknown protocol message

Suricata-verify tests: https://github.com/OISF/suricata-verify/pull/28

### CODE REVIEW

File: src/app-layer-ftp.c [hunk 1]
```diff
@@ -374,6 +378,55 @@ static void FtpTransferCmdFree(void *data)
...
+}
+
+/**
+ * \brief This function is extract a port number from the command input line for IPv4 FTP usage
+ * \param input input line of the command
+ * \param input_len length of the request
+ *
+ * \retval 0 if a port number could not be extracted; otherwise, thed ynamic port number
+ */
+static int FTPGetV4PortNumber(uint8_t *input, uint32_t input_len)
```
Comment (victorjulien) [inline src/app-layer-ftp.c]: should we have validation in place on: numbers non-negative, values not leading to overflow, no missing values?

File: src/app-layer-ftp.c [hunk 2]
```diff
@@ -552,6 +597,29 @@ static int FTPParseResponse(Flow *f, void *ftp_state, AppLayerParserState *pstat
         }
     }
 
+    if (state->command == FTP_COMMAND_EPRT) {
+        uint16_t dyn_port;
+        dyn_port = FTPGetV6PortNumber(state->port_line, state->port_line_len);
```
Comment (victorjulien) [inline src/app-layer-ftp.c]: declare var on this line itself for more compactness

File: src/app-layer-ftp.c [hunk 3]
```diff
@@ -552,6 +597,29 @@ static int FTPParseResponse(Flow *f, void *ftp_state, AppLayerParserState *pstat
...
+        }
+        state->dyn_port = dyn_port;
+        state->active = 1;
+        SCLogDebug("FTP active mode (v6): dynamic port %"PRIu16"", dyn_port);
+    }
+
+    if (state->command == FTP_COMMAND_PORT) {
+        if ((flags & STREAM_TOCLIENT)) {
+            uint16_t dyn_port;
+            dyn_port = FTPGetV4PortNumber(state->port_line, state->port_line_len);
```
Comment (victorjulien) [inline src/app-layer-ftp.c]: declare var here

File: src/app-layer-ftp.c [hunk 4]
```diff
@@ -552,6 +597,29 @@ static int FTPParseResponse(Flow *f, void *ftp_state, AppLayerParserState *pstat
...
+        if (dyn_port == 0) {
+            return 0;
+        }
+        state->dyn_port = dyn_port;
+        state->active = 1;
+        SCLogDebug("FTP active mode (v6): dynamic port %"PRIu16"", dyn_port);
+    }
+
+    if (state->command == FTP_COMMAND_PORT) {
+        if ((flags & STREAM_TOCLIENT)) {
```
Comment (victorjulien) [inline src/app-layer-ftp.c]: what happens if direction is toserver?

File: src/app-layer-ftp.h [hunk 5]
```diff
@@ -115,6 +116,7 @@ typedef struct FtpState_ {
     uint8_t *input;
     int32_t input_len;
     uint8_t direction;
+    uint8_t active;
```
Comment (victorjulien) [inline src/app-layer-ftp.h]: use 'bool' instead? Plus then true/false instead of 1/0

File: src/app-layer-ftp.c (hunk 1 above)
Comment (jlucovsky) [reply src/app-layer-ftp.c]: The port number is  an unsigned 16-bite value -- all values other than 0 are valid. I'm not following the "no missing values" as a single value, not a range or other sequence, is being parsed.

File: src/app-layer-ftp.c (hunk 2 above)
Comment (jlucovsky) [reply src/app-layer-ftp.c]: Will do.

File: src/app-layer-ftp.c (hunk 3 above)
Comment (jlucovsky) [reply src/app-layer-ftp.c]: will do.

File: src/app-layer-ftp.c (hunk 4 above)
Comment (jlucovsky) [reply src/app-layer-ftp.c]: No action is necessary -- handling for `EPSV` and `PASV` will parse the parse numbers with those command lines.

File: src/app-layer-ftp.h (hunk 5 above)
Comment (jlucovsky) [reply src/app-layer-ftp.h]: Will do.

**Comment by jlucovsky** [comment]
Continued in #3768

File: src/app-layer-ftp.c (hunk 1 above)
Comment (victorjulien) [reply src/app-layer-ftp.c]: My point is that it makes sense to have validation that the values are valid. If the command badly formatted we should set an event and not create the expectation.

**Pull request closed without merging**